ostd/src/mm/io.rs

```diff
@@ -672,3 +730,26 @@ impl<'a> From<&'a mut [u8]> for VmWriter<'a> {
         unsafe { Self::from_kernel_space(slice.as_mut_ptr(), slice.len()) }
     }
 }
+
+/// Asserts that reading or writing a `T` can be implemented with a single instruction.
+///
+/// We currently rely on this assertion to ensure that the memory operation created by
+/// `ptr::read_volatile` and `ptr::write_volatile` doesn't tear. However, the Rust documentation
+/// makes no such guarantee, and even the wording in the LLVM LangRef is ambiguous.
+///
+/// At this point, we can only _hope_ that this doesn't break in future versions of the Rust or
+/// LLVM compilers. However, this is unlikely to happen in practice, since the Linux kernel also
+/// uses "volatile" semantics to implement `READ_ONCE`/`WRITE_ONCE`.
+#[cfg(target_arch = "x86_64")]
+const fn assert_memory_access_no_tearing<T: Pod>() {
+    let size = size_of::<T>();
+
+    if size == 0 || size == 1 || size == 2 || size == 4 || size == 8 {
```

Comment:
I don't think we want to read zero-sized values with `DmaCoherent`. Such usage almost always indicates a bug.